File: leetcode js/283. Move Zeroes.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
/**
 * 思路：
 * 先找出所有的0，并将位置存储在一个数组中（因为一边遍历数组一边直接用splice删除是大忌，所以不这样，因为位置是即时改变的）
 * 然后倒叙删除相应位置的0，因为免得先删除前边，数组后边0的位置改变
 * 在往数组push几个0
 */
var moveZeroes = function(nums) {

    var len = nums.length,
        array = [],
        zeroNums = 0,
        i;

    for(i = 0;i < len;i++) {
        if(nums[i] === 0) {
            array[zeroNums++] = i;
        }
    }

    for(i = zeroNums - 1;i >= 0;i--) {
        nums.splice(array[i],1);
    }

    for(i = 0;i < zeroNums;i++) {
        nums.push(0);
    }
};

File: leetcode java/Array/283. Move Zeroes.java
/*
Given an array nums, write a function to move all 0's to the end of it
while maintaining the relative order of the non-zero elements.

For example, given nums = [0, 1, 0, 3, 12], after calling your function,
nums should be [1, 3, 12, 0, 0].
Note:

1.You must do this in-place without making a copy of the array.
2.Minimize the total number of operations.

 */
/*
思路：
使用两个指针，一个遍历，一个方便存储
 */
 public class Solution {
    public void moveZeroes(int[] nums) {
        int len = nums.length,
            i = 0,
            j = 0;

        for(;i < len;i++) {
            if(nums[i] != 0) {
                if(i != j) {
                    nums[j] = nums[i];
                    nums[i] = 0;
                }
                j++;
            }
        }
    }
}
